File: leetcode/Python/verify-preorder-serialization-of-a-binary-tree.py
class Solution:
    def __init__(self):
        self.v = None
        self.idx = 0
    def isValidPreorder(self):
        check_size = True if self.idx==0 else False
        if self.idx >= len(self.v):
            return False
        elif self.v[self.idx]=='#':
            return False if check_size and self.idx+1<len(self.v) else True
        self.idx+=1
        if not self.isValidPreorder():
            return False
        self.idx+=1
        if not self.isValidPreorder():
            return False
        return not check_size or (check_size and self.idx+1==len(self.v))
    def isValidSerialization(self, preorder: str) -> bool:
        self.v = preorder.split(',')
        return self.isValidPreorder()

File: leetcode/C++/verify-preorder-serialization-of-a-binary-tree.cpp
class Solution {
public:
    vector<string> v;
    void parsing(string &preorder){
        istringstream ss(preorder);
        string buf;
        while (getline(ss, buf, ',')) v.push_back(buf);
    }
    bool isValidPreorder(int &idx){
        bool size_check = (idx==0)? true : false;
        if(idx>=v.size()) return false;
        else if(v[idx]=="#") {
            if(size_check && idx+1!=v.size()) return false;
            return true;
        }
        return isValidPreorder(++idx)&&isValidPreorder(++idx)&&(!size_check || size_check&&v.size()==idx+1);
    }
    bool isValidSerialization(string preorder) {
        int idx=0;
        parsing(preorder);
        return isValidPreorder(idx);
    }
};